Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,239 @@
+        if (rads <= MinIntensity)
+            return null;
+
+        // if source and destination on the same grid it's possible that
+        // between them can be another grid (ie. shuttle in center of donut station)
+        // however we can do simplification and ignore that case
+        if (GridcastSimplifiedSameGrid && sourceTrs.GridUid != null && sourceTrs.GridUid == destTrs.GridUid)
+        {
+            return Gridcast(mapId, sourceTrs.GridUid.Value, sourceUid, destUid,
+                sourceWorld, destWorld, rads, resistanceQuery);
+        }
+
+        // lets check how many grids are between source and destination
+        // do a box intersection test between target and destination
+        // it's not very precise, but really cheap
+        var box = Box2.FromTwoPoints(sourceWorld, destWorld);
+        var grids = _mapManager.FindGridsIntersecting(mapId, box, true);
+
+        // we are only interested in grids that has some radiation blockers
+        var resGrids = grids.Where(grid => resistanceQuery.HasComponent(grid.GridEntityId)).ToArray();
+        var resGridsCount = resGrids.Length;
+
+        if (resGridsCount == 0)
+        {
+            // no grids found - so no blockers (just distance penalty)
+            return new RadiationRay(mapId, sourceUid,sourceWorld,
+                destUid,destWorld, rads);
+        }
+        else if (resGridsCount == 1)
+        {
+            // one grid found - use it for gridcast
+            return Gridcast(mapId, resGrids[0].GridEntityId, sourceUid, destUid,
+                sourceWorld, destWorld, rads, resistanceQuery);
+        }
+        else
+        {
+            // more than one grid - fallback to raycast
+            return Raycast(mapId, sourceUid, destUid, sourceWorld, destWorld,
+                dir.Normalized, dist, rads, blockerQuery);
+        }
+    }
+
+    private RadiationRay Gridcast(MapId mapId, EntityUid gridUid, EntityUid sourceUid, EntityUid destUid,
+        Vector2 sourceWorld, Vector2 destWorld, float incomingRads,
+        EntityQuery<RadiationGridResistanceComponent> resistanceQuery)
+    {
+        var visitedTiles = new List<(Vector2i, float?)>();
+        var radRay = new RadiationRay(mapId, sourceUid,sourceWorld,
+            destUid,destWorld, incomingRads)
+        {
+            Grid = gridUid,
+            VisitedTiles = visitedTiles
+        };
+
+        // if grid doesn't have resistance map just apply distance penalty
+        if (!resistanceQuery.TryGetComponent(gridUid, out var resistance))
+            return radRay;
+        var resistanceMap = resistance.ResistancePerTile;
+
+        // get coordinate of source and destination in grid coordinates
+        // todo: entity queries doesn't support interface - use it when IMapGridComponent will be removed
+        if (!TryComp(gridUid, out IMapGridComponent? grid))
+            return radRay;
+        var sourceGrid = grid.Grid.TileIndicesFor(sourceWorld);
```

Comment:
`TileIndicesFor()` uses `WorldToLocal()` which in turn involves getting the grid's transform component & computing the `InvWorldMatrix`. If you have a big radiation source on the grid, it'd be much better to just compute the inv world matrix once and cache it, rather than re-getting the component & re-compute it this for each source.

Alternatively:
I Imagine that the majority of the passive radiation stuff will all be same grid stuff where the source & target are both directly parented to the grid. So an easier option to speed this up w/o figuring out how to best cache the inv world matrix might be to just use the entity coordinates rather than world coordinates for the source & target, because if `parentUid == gridUid` (i.e., directly parented to the grid), then it just directly uses local coordinates and doesn't have to do any map -> grid conversions.

Reply:
>If you have a big radiation source on the grid, it'd be much better to just compute the inv world matrix once and cache it, rather than re-getting the component & re-compute it this for each source.

Imo, I would cache inv transform matrix in `IMapGrid` itself. I think it might be reused by other systems. Adding it here will just make code harder to read.

> I Imagine that the majority of the passive radiation stuff will all be same grid stuff where the source & target are both directly parented to the grid.

Yeah, thats true for some cases. But if radiation source is placed in space near station, it won't work. Adding another snowflake scenario looks like overkill for me.